miniapps/autodiff/adnonlininteg.hpp

```diff
@@ -0,0 +1,503 @@
+/// energy and residual at a point, provides the derivative of the residual
+/// with respect to the active arguments of the functor.
+///
+/**
+   The two templated parameters are
+   template<typename, typename> class CTD and int m.
+   The templated class CTD is a user-provided functor which implements
+   the function of interest. The CTD class should have the following signature:
+
+template<typename DType, typename MVType>
+
+class MyQFunctorJ
+
+{
+
+    DType operator()(const Vector &vparam, MVType &uu)
+
+        {...}
+
+    void operator()(const Vector &vparam, MVType &uu, MVType &rr)
+
+    {...}
+
+};
+
+
+DType - represents the scalars in the functor, and  MVType - represents
+the vectors in the functor. In ADQFunctionTJ,  DType will be replaced
+either with double or AD-type.
+The first operator()(const Vector &vparam, MVType &uu) takes as input
+standard MFEM Vector vparam (passive arguments), which holds all parameters
+ supplied to the functor and MVType vector uu, which holds all active
+arguments provided to the functor. The operator returns a scalar
+representing the value of the function(energy) for passive parameters vparam
+and active arguments uu.
+
+
+The second void operator()(const Vector &vparam, MVType &uu, MVType &rr)
+returns a vector-valued function rr with passive input arguments vparam
+and active (AD) arguments supplied in vector uu. Since the size of the
+return vector rr is not known in advance, it should be provided explicitly
+to the ADQFunctionTJ class, i.e., the template parameter m in ADQFunctionTJ
+represents the dimensions of the return vector rr.
+
+
+For PDE discretized problem, the first operator should return the energy
+evaluated a point, and the second operator should return the residual
+evaluated at the point. The derivative of the residual vector rr with
+respect to the active arguments uu is evaluated automatically by
+ADQFunctionTJ and returned by the ADQFunctionTJ:: QFunctionDD(..) method.
+*/
+template<template<typename, typename> class CTD, int m>
+class ADQFunctionTJ
+{
+   // m - dimension of the residual vector
+   // the Jacobian will have dimensions [m,length(uu)]
+protected:
+#ifdef MFEM_USE_ADEPT
+   adept::Stack m_stack;
+#endif
+
+public:
+#if defined MFEM_USE_ADEPT
+   /// AD-type based on the Adept AD library.
+   typedef adept::adouble ADFType;
+   typedef TADVector<ADFType> ADFVector;
+   typedef TADDenseMatrix<ADFType> ADFDenseMatrix;
+#elif defined MFEM_USE_FADBADPP
+#ifdef MFEM_USE_ADFORWARD
+   /// AD-type based on the FADBAD++ library -
+   /// used only for forward differentiation.
+   typedef fadbad::F<double> ADFType;
+   typedef TADVector<ADFType> ADFVector;
+   typedef TADDenseMatrix<ADFType> ADFDenseMatrix;
+#else
+   /// AD-type based on the FADBAD++ library -
+   /// used only for reverse AD mode.
+   typedef fadbad::B<double> ADFType;
+   typedef TADVector<ADFType> ADFVector;
+   typedef TADDenseMatrix<ADFType> ADFDenseMatrix;
+#endif
```

Comment:
These libraries are not supported in the PR

Reply:
This file is left over from the previous PR - it is removed.